pkcs11/_pkcs11.pyx

```diff
@@ -1084,414 +1346,445 @@ class DomainParameters(types.DomainParameters):
+        return bytes(self.output_buf[:self.buffer_data_length])
 
-            ciphertext = CK_BYTE_buffer(length)
+    cdef CK_RV update_resizing_output(
+            self,
+            KeyOperationUpdateWithResult op_update,
+            CK_BYTE *data,
+            CK_ULONG data_len,
+    ) with gil:
+
+        cdef CK_ULONG length = self.buffer_size
+        cdef CK_BYTE *output_buf_loc = &self.output_buf[0]
+        cdef CK_RV retval
 
+        with nogil:
+            retval = op_update(
+                self.session.handle, data, data_len, output_buf_loc, &length
+            )
+
+        if retval == CKR_BUFFER_TOO_SMALL:
+            self.resize_buffer(length)
+            output_buf_loc = &self.output_buf[0]
             with nogil:
-                retval = _funclist.C_Encrypt(handle, data_ptr, data_len, &ciphertext[0], &length)
-            assertRV(retval)
+                retval = op_update(
+                    self.session.handle, data, data_len, output_buf_loc, &length
+                )
+        self.buffer_data_length = length
+        return retval
 
-            return bytes(ciphertext[:length])
+    cdef CK_RV execute_resizing_output(self, KeyOperationWithResult op) with gil:
 
+        cdef CK_ULONG length = self.buffer_size
+        cdef CK_BYTE *output_buf_loc = &self.output_buf[0]
+        cdef CK_RV retval
 
-    def _encrypt_generator(self, data,
-                           mechanism=None, mechanism_param=None,
-                           buffer_size=8192):
-        """
-        Do chunked encryption.
+        with nogil:
+            retval = op(self.session.handle, output_buf_loc, &length)
 
-        Failing to consume the generator will raise GeneratorExit when it
-        garbage collects. This will release the lock, but you'll still be
-        in the middle of an operation, and all future operations will raise
-        OperationActive, see tests/test_iterators.py:test_close_iterators().
+        if retval == CKR_BUFFER_TOO_SMALL:
+            self.resize_buffer(length)
+            output_buf_loc = &self.output_buf[0]
+            with nogil:
+                retval = op(self.session.handle, output_buf_loc, &length)
 
-        FIXME: cancel the operation when we exit the generator early.
-        """
-        mech = MechanismWithParam(
-            self.key_type, DEFAULT_ENCRYPT_MECHANISMS,
-            mechanism, mechanism_param)
+        self.buffer_data_length = length
+        return retval
 
-        cdef CK_SESSION_HANDLE handle = self.session._handle
-        cdef CK_MECHANISM *mech_data = mech.data
-        cdef CK_OBJECT_HANDLE key = self._handle
+    def _cancel_operation(self, silent):
+        cdef CK_RV retval
+        if self.session.token.slot.cryptoki_version >= (3, 0):
+            # cancel the operation if still active
+            # This is a PKCS#11 3.x feature
+            with nogil:
+                retval = self.op_init(self.session.handle, NULL, self.key)
```

Comment:
It's a bit of a shame that this very elegant way of cancelling operations is only available in modules that support version 3.x...